server/src/main/java/org/elasticsearch/cluster/ClusterStateTaskExecutor.java

```diff
@@ -202,4 +202,75 @@ public Releasable dropHeadersContext() {
             return dropHeadersContextSupplier.get();
         }
     }
+
+    /**
+     * A basic implementation for batch executors that simply need to execute the tasks in the batch iteratively,
+     * producing a cluster state after each task. This allows executing the tasks in the batch as a
+     * series of executions, each taking an input cluster state and producing a new cluster state that serves as the
+     * input of the next task in the batch.
+     */
+    abstract class DefaultBatchExecutor<T extends ClusterStateTaskListener> implements ClusterStateTaskExecutor<T> {
+
+        /**
+         * Executes the provided task from the batch.
+         *
+         * @param taskContext The task to be executed.
+         * @param curState The cluster state on which the task should be executed.
+         * @return The resulting cluster state after executing this task.
+         */
+        public abstract ClusterState executeTask(TaskContext<T> taskContext, ClusterState curState);
+
+        /**
+         * Called once all tasks in the batch have finished execution. It should return a cluster state that reflects
+         * the execution of all the tasks.
+         *
+         * @param batchExecutionContext The tasks that were executed.
+         * @param initState The cluster state before executing the tasks.
+         * @param curState The cluster state resulting from the execution of all the tasks.
+         * @return The resulting cluster state after executing all the tasks.
+         */
+        public ClusterState afterBatchExecution(
+            BatchExecutionContext<T> batchExecutionContext,
+            ClusterState initState,
```

Comment:
I feel bad about giving folks access to the initial state here. I think that could lead to some unexpected behaviour. Really the only question should be whether there were any changes. WDYT about having two methods, one called if `initState == curState` and the other called if something changed? Either that or a boolean parameter.

Reply:
I think the flag would be good since then the afterBatchExecution logic would be all in one place. If we take away the initState, [DeleteDesiredNodesExecutor](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/admin/cluster/desirednodes/TransportDeleteDesiredNodesAction.java#L82) wouldn't fit into this. I guess it is a special "low-level" executor anyway.

Reply:
so instead of 
```
initialState().copyAndUpdateMetadata(metadata -> metadata.removeCustom(DesiredNodesMetadata.TYPE))
```
I can just use 
```
clusterState.copyAndUpdateMetadata(metadata -> metadata.removeCustom(DesiredNodesMetadata.TYPE))
```
and they'd be the same?

Reply:
Right.

Although that does raise a good point that `BatchExecutionContext` gives access to the initial state. And the tasks. I think we don't want that either, we should always run this method within a `dropHeadersContext()` and then it should only need the state.